// hazelcast-client/lib/proxy/BaseProxy.js
"use strict";
/*
 * Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/** @ignore */ /** */
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseProxy = void 0;
const BuildInfo_1 = require("../BuildInfo");
/**
 * Common super class for any proxy.
 * @internal
 */
class BaseProxy {
    constructor(serviceName, name, proxyManager, partitionService, invocationService, serializationService, listenerService, clusterService, connectionRegistry) {
        this.serviceName = serviceName;
        this.name = name;
        this.proxyManager = proxyManager;
        this.partitionService = partitionService;
        this.invocationService = invocationService;
        this.serializationService = serializationService;
        this.listenerService = listenerService;
        this.clusterService = clusterService;
        this.connectionRegistry = connectionRegistry;
    }
    getPartitionKey() {
        return this.name;
    }
    getName() {
        return this.name;
    }
    getServiceName() {
        return this.serviceName;
    }
    destroy() {
        return this.proxyManager.destroyProxy(this.name, this.serviceName).then(() => {
            return this.postDestroy();
        });
    }
    destroyLocally() {
        return this.postDestroy();
    }
    postDestroy() {
        return Promise.resolve();
    }
    /**
     * Encodes a request from a codec and invokes it on owner node of given key.
     */
    encodeInvokeOnKey(codec, partitionKey, ...codecArguments) {
        const partitionId = this.partitionService.getPartitionId(partitionKey);
        return this.encodeInvokeOnPartition(codec, partitionId, ...codecArguments);
    }
    /**
     * Encodes a request from a codec and invokes it on owner node of given key.
     * This method also overrides invocation timeout.
     */
    encodeInvokeOnKeyWithTimeout(timeoutMillis, codec, partitionKey, ...codecArguments) {
        const partitionId = this.partitionService.getPartitionId(partitionKey);
        return this.encodeInvokeOnPartitionWithTimeout(timeoutMillis, codec, partitionId, ...codecArguments);
    }
    /**
     * Encodes a request from a codec and invokes it on any node.
     */
    encodeInvokeOnRandomTarget(codec, ...codecArguments) {
        const clientMessage = codec.encodeRequest(this.name, ...codecArguments);
        return this.invocationService.invokeOnRandomTarget(clientMessage);
    }
    encodeInvokeOnTarget(codec, target, ...codecArguments) {
        const clientMessage = codec.encodeRequest(this.name, ...codecArguments);
        return this.invocationService.invokeOnTarget(clientMessage, target);
    }
    /**
     * Encodes a request from a codec and invokes it on owner node of given partition.
     */
    encodeInvokeOnPartition(codec, partitionId, ...codecArguments) {
        const clientMessage = codec.encodeRequest(this.name, ...codecArguments);
        return this.invocationService.invokeOnPartition(clientMessage, partitionId);
    }
    /**
     * Encodes a request from a codec and invokes it on owner node of given partition.
     * This method also overrides invocation timeout.
     */
    encodeInvokeOnPartitionWithTimeout(timeoutMillis, codec, partitionId, ...codecArguments) {
        const clientMessage = codec.encodeRequest(this.name, ...codecArguments);
        return this.invocationService.invokeOnPartition(clientMessage, partitionId, timeoutMillis);
    }
    /**
     * Serializes an object according to serialization settings of the client.
     */
    toData(object) {
        return this.serializationService.toData(object);
    }
    /**
     * De-serializes an object from binary form according to serialization settings of the client.
     */
    toObject(data) {
        return this.serializationService.toObject(data);
    }
    getConnectedServerVersion() {
        const activeConnections = this.connectionRegistry.getConnections();
        for (const address in activeConnections) {
            return activeConnections[address].getConnectedServerVersion();
        }
        return BuildInfo_1.BuildInfo.UNKNOWN_VERSION_ID;
    }
}
exports.BaseProxy = BaseProxy;
